File: app/src/main/java/de/westnordost/streetcomplete/FragmentContainerActivity.java
package de.westnordost.streetcomplete;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

public class FragmentContainerActivity extends AppCompatActivity
{
	public static final String EXTRA_FRAGMENT_CLASS = "de.westnordost.streetcomplete.fragment_class";

	@Override protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_fragment_container);

		Toolbar toolbar = findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);
		getSupportActionBar().setDisplayShowHomeEnabled(true);
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
	}

	@Override protected void onPostCreate(@Nullable Bundle savedInstanceState)
	{
		super.onPostCreate(savedInstanceState);
		if(savedInstanceState == null)
		{
			String fragmentClass = getIntent().getStringExtra(EXTRA_FRAGMENT_CLASS);
			if (fragmentClass != null)
			{
				try
				{
					Class c = Class.forName(fragmentClass);
					Fragment f = (Fragment) c.newInstance();
					f.setArguments(getIntent().getExtras());
					setCurrentFragment(f, false);
				} catch (Exception e)
				{
					throw new RuntimeException(e);
				}
			}
		}
	}

	public void setCurrentFragment(Fragment fragment)
	{
		setCurrentFragment(fragment, true);
	}

	private void setCurrentFragment(Fragment fragment, boolean addToBackStack)
	{
		FragmentTransaction tr = getSupportFragmentManager().beginTransaction();
		tr.setCustomAnimations(
				R.anim.enter_from_right, R.anim.exit_to_left,
				R.anim.enter_from_left, R.anim.exit_to_right);
		tr.replace(R.id.fragment_container, fragment);
		if(addToBackStack) tr.addToBackStack(null);
		tr.commit();
	}

	private Fragment getCurrentFragment()
	{
		return getSupportFragmentManager().findFragmentById(R.id.fragment_container);
	}

	@Override public void onBackPressed() {
		if (getSupportFragmentManager().getBackStackEntryCount() > 0) {
			getSupportFragmentManager().popBackStack();
		} else {
			super.onBackPressed();
		}
	}

	@Override public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case android.R.id.home:
				onBackPressed();
				return true;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override protected void onNewIntent(Intent intent)
	{
		super.onNewIntent(intent);

		Fragment f = getCurrentFragment();
		if(f instanceof IntentListener)
		{
			((IntentListener)f).onNewIntent(intent);
		}
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/FragmentContainerActivity.kt
package de.westnordost.streetcomplete

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentManager.POP_BACK_STACK_INCLUSIVE
import androidx.fragment.app.commit

/** An activity that contains one full-screen ("main") fragment */
open class FragmentContainerActivity(
    @LayoutRes contentLayoutId: Int = R.layout.activity_fragment_container
) : AppCompatActivity(contentLayoutId), DisplaysTitle {

    var mainFragment: Fragment?
        get() = supportFragmentManager.findFragmentById(R.id.fragment_container)
        set(value) {
            supportFragmentManager.popBackStack("main", POP_BACK_STACK_INCLUSIVE)
            if (value != null) {
                supportFragmentManager.commit { replace(R.id.fragment_container, value) }
            }
        }

    private val fragmentLifecycleCallbacks = object : FragmentManager.FragmentLifecycleCallbacks() {
        override fun onFragmentStarted(fragmentManager: FragmentManager, fragment: Fragment) {
            if (fragment.id == R.id.fragment_container && fragment is HasTitle) {
                updateTitle(fragment)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        if (toolbar != null) {
            setSupportActionBar(toolbar)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }
        supportFragmentManager.registerFragmentLifecycleCallbacks(fragmentLifecycleCallbacks, false)
    }

    fun pushMainFragment(fragment: Fragment) {
        supportFragmentManager.commit {
            setCustomAnimations(
                R.anim.enter_from_right, R.anim.exit_to_left,
                R.anim.enter_from_left, R.anim.exit_to_right
            )
            replace(R.id.fragment_container, fragment)
            addToBackStack("main")
        }
    }

    override fun updateTitle(fragment: HasTitle) {
        title = fragment.title
        supportActionBar?.subtitle = fragment.subtitle
    }

    override fun onBackPressed() {
        if ((mainFragment as? BackPressedListener)?.onBackPressed() == true) {
            return
        }
        if (supportFragmentManager.backStackEntryCount > 0) {
            supportFragmentManager.popBackStack()
            return
        }
        super.onBackPressed()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        } else super.onOptionsItemSelected(item)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        (mainFragment as? IntentListener)?.onNewIntent(intent)
    }
}
